### PR TITLE
ARROW-6967: [C++][Dataset] IN, IS_VALID filter expressions

### PR DESCRIPTION
**Pull request by bkietz**
Add expressions for filtering on membership in a set (IsIn kernel), validity of an expression (by wrapping the null bitmap into a BooleanArray)

### CODE REVIEW

File: cpp/src/arrow/dataset/filter.cc [hunk 1]
```diff
@@ -220,6 +221,23 @@ std::shared_ptr<Expression> Invert(const Expression& expr) {
...
+std::shared_ptr<Expression> Expression::Assume(const Expression& given) const {
+  if (given.type() == ExpressionType::COMPARISON) {
+    const auto& cmp = checked_cast<const ComparisonExpression&>(given);
+    if (cmp.op() == compute::CompareOperator::EQUAL) {
+      if (this->Equals(cmp.left_operand())) {
+        return cmp.right_operand();
+      }
+
+      if (this->Equals(cmp.right_operand())) {
+        return cmp.right_operand();
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.cc]: `left_operand` perhaps?

File: cpp/src/arrow/dataset/filter.cc [hunk 2]
```diff
@@ -220,6 +221,23 @@ std::shared_ptr<Expression> Invert(const Expression& expr) {
   return nullptr;
 }
 
+std::shared_ptr<Expression> Expression::Assume(const Expression& given) const {
+  if (given.type() == ExpressionType::COMPARISON) {
+    const auto& cmp = checked_cast<const ComparisonExpression&>(given);
+    if (cmp.op() == compute::CompareOperator::EQUAL) {
+      if (this->Equals(cmp.left_operand())) {
+        return cmp.right_operand();
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.cc]: I'm not sure of the point of this. It seems to be assuming that `right_operand` is somehow simpler (and therefore more desirable) than `this`. But is it the case generally?

Perhaps this should be restricted to constant comparands...

File: cpp/src/arrow/dataset/filter.h [hunk 3]
```diff
@@ -305,6 +310,37 @@ class ARROW_DS_EXPORT NotExpression final
...
+  InExpression(std::shared_ptr<Array> set, std::shared_ptr<Expression> operand)
+      : ExpressionImpl(std::move(operand)), set_(std::move(set)) {}
+
+  std::string ToString() const override;
+
+  Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
+
+  std::shared_ptr<Expression> Assume(const Expression& given) const override;
+
+  const std::shared_ptr<Array>& set() const { return set_; }
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.h]: Add a docstring?

File: cpp/src/arrow/dataset/filter.h [hunk 4]
```diff
@@ -305,6 +310,37 @@ class ARROW_DS_EXPORT NotExpression final
...
+
+  Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
+
+  std::shared_ptr<Expression> Assume(const Expression& given) const override;
+};
+
+class ARROW_DS_EXPORT InExpression final
+    : public ExpressionImpl<UnaryExpression, InExpression, ExpressionType::IN> {
+ public:
+  InExpression(std::shared_ptr<Array> set, std::shared_ptr<Expression> operand)
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.h]: Not important, but if it's "operand IN set", then perhaps `(operand, set)` would be a more natural argument order.

File: cpp/src/arrow/dataset/filter_test.cc [hunk 5]
```diff
@@ -108,6 +108,14 @@ TEST_F(ExpressionsTest, SimplificationOfCompoundQuery) {
   AssertSimplifiesTo("b"_ == 4, "a"_ == 0, "b"_ == 4);
 
   AssertSimplifiesTo("a"_ == 3 or "b"_ == 4, "a"_ == 0, "b"_ == 4);
+
+  // TODO(bkietz) the following simplifications should be possible:
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter_test.cc]: Which ones? This test fails?

File: cpp/src/arrow/dataset/filter.cc [hunk 6]
```diff
@@ -882,6 +963,36 @@ Result<Datum> TreeEvaluator::Evaluate(const NotExpression& expr,
...
+    return Datum(false);
+  }
+
+  if (operand_values.is_scalar()) {
+    return Datum(true);
+  }
+
+  DCHECK(operand_values.is_array());
+  return Datum(std::make_shared<BooleanArray>(operand_values.array()->length,
+                                              operand_values.array()->buffers[0]));
```
Comment (pitrou) [inline cpp/src/arrow/dataset/filter.cc]: This may construct an array with null data, if the operand has no nulls.

**Review by pitrou — COMMENTED** [review]
Some questions here.

File: cpp/src/arrow/dataset/filter.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.cc]: yes, thanks

File: cpp/src/arrow/dataset/filter.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.cc]: The point is to allow `"beta"_.Assume("beta"_ == 3).Equals(3)`,
which in turn allows `"beta"_.In({1, 2, 3}).Assume("beta"_ == 0).Equals(false)`.

Restricting it to constant given.right_operand seems sensible

File: cpp/src/arrow/dataset/filter.h (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.h]: will do

File: cpp/src/arrow/dataset/filter.h (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.h]: will do

File: cpp/src/arrow/dataset/filter.cc (hunk 6 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.cc]: thanks, I'll add a check for null_count == 0

**Review by nealrichardson — APPROVED** [review]
I didn't wade into the implementation details but the interface and test expectations seem reasonable enough to me.

You should revise the Jira title to match the scope here and make any followups needed (i.e. it talks about COALESCE but that's not in here AFAICT).

**Comment by bkietz** [comment]
I've amended the JIRA and added a follow up for COALESCE https://issues.apache.org/jira/browse/ARROW-7179

**Pull request closed without merging**